### PR TITLE
Add a clusterEviction plugin

### PR DESCRIPTION
**Pull request by Poor12**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add a clusterEviction plugin to filter cluster which has been evicted.

**Which issue(s) this PR fixes**:
Part of #3169 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```
karmada-scheduler: add a clusterEviction plugin.
```

### CODE REVIEW

**Comment by jwcesign** [comment]
cc @XiShanYongYe-Chang for checking

File: test/e2e/failover_test.go [hunk 1]
```diff
@@ -141,6 +142,106 @@ var _ = framework.SerialDescribe("failover testing", func() {
...
 	})
+
+	ginkgo.Context("Taint cluster testing", func() {
+		var policyNamespace, policyName string
+		var deploymentNamespace, deploymentName string
+		var deployment *appsv1.Deployment
+		var taint corev1.Taint
+		var maxGroups, minGroups, numOfFailedClusters int
+		var policy *policyv1alpha1.PropagationPolicy
+		maxGroups = 1
```
Comment (jwcesign) [inline test/e2e/failover_test.go:153]: maxGroups, minGroups, numOfFailedClusters := 1,1,1

**Review by jwcesign — COMMENTED** [review]
other  lgtm

File: test/e2e/failover_test.go (hunk 1 above)
Comment (Poor12) [reply test/e2e/failover_test.go:153]: /cc @XiShanYongYe-Chang. I follow the former code style. To be honest, I do not know which style is better.

Comment (XiShanYongYe-Chang) [reply test/e2e/failover_test.go:153]: It's just a code style, any writing is OK. Keep the overall consistency will be nice.

File: pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	return &ClusterEviction{}, nil
+}
+
+// Name returns the plugin name.
+func (p *ClusterEviction) Name() string {
+	return Name
+}
+
+// Filter checks if the target cluster is in the GracefulEvictionTasks which means it has been evicted.
+func (p *ClusterEviction) Filter(_ context.Context, bindingSpec *workv1alpha2.ResourceBindingSpec, _ *workv1alpha2.ResourceBindingStatus, cluster *clusterv1alpha1.Cluster) *framework.Result {
```
Comment (chaunceyjiang) [inline pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:35]: Does this plugin only take effect when the `Application Failover` feature is enabled?

Comment (Poor12) [reply pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:35]: Not only application failover, but also the current cluster failover. See comments: https://github.com/karmada-io/karmada/pull/3456#issuecomment-1522660388.

Comment (chaunceyjiang) [reply pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:35]: Does this conflict with the `TaintToleration` plugin?

Comment (jwcesign) [reply pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:35]: I don't think there will be any conflict. The plugins will execute sequentially, ensuring that the final result does not include the cluster in eviction tasks, as intended.

Comment (chaunceyjiang) [reply pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:35]: ok.

File: pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/karmada-io/karmada/pkg/scheduler/framework"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+const (
+	// Name is the name of the plugin used in the plugin registry and configurations.
+	Name = "ClusterEviction"
+)
+
+// ClusterEviction is a plugin that checks if the target cluster is in the GracefulEvictionTasks which means it has been evicted.
```
Comment (RainbowMango) [inline pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:19]: A cluster in eviction tasks doesn't mean it `has been` evicted, it means it is in the process of eviction.

File: pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+type ClusterEviction struct{}
+
+var _ framework.FilterPlugin = &ClusterEviction{}
+
+// New instantiates the APIEnablement plugin.
```
Comment (RainbowMango) [inline pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go:24]: `APIEnablement`

Comment (Poor12) [reply pkg/scheduler/framework/plugins/clustereviction/cluster_eviction.go]: Done.

**Review by RainbowMango — APPROVED** [review]
@jwcesign @chaunceyjiang Do you have any other comments? Can we move forward now?

**Pull request merged**